tests/conftest.py

```diff
@@ -64,3 +82,79 @@ def pytest_collection_modifyitems(config, items):
         item.add_marker(
             pytest.mark.filterwarnings("error::pydantic.PydanticDeprecatedSince20")
         )
+
+
+def is_available(
+    ctr_client: DockerClient,
+) -> Union[Tuple[Literal[True], None], Tuple[Literal[False], str]]:
+    """Returns a tuple (True, None) if the container runtime is available
+    and (False, reason (str)) if it's not available."""
+
+    try:
+        # TODO: Implement 'DockerClient.version' and use that instead.
+        subprocess.run(
+            ctr_client.client_config.client_call + ["version"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            text=True,
+            check=True,
+            timeout=10,
+        )
+        return True, None
+    except (
+        subprocess.CalledProcessError,
+        subprocess.TimeoutExpired,
+        FileNotFoundError,
+    ) as exc:
+        if isinstance(exc, subprocess.CalledProcessError):
+            reason = "\n" + exc.output
+        elif isinstance(exc, subprocess.TimeoutExpired):
+            reason = f"timed out after {exc.timeout} seconds"
+        elif isinstance(exc, FileNotFoundError):
+            reason = "executable not found"
+        command = " ".join(ctr_client.client_config.client_call)
+        return False, (
+            f"Unable to get version with command"
+            f" '{command} version'. Reason: {reason}"
+        )
+
+
+class TestSessionClient:

```

Comment:
It feels like this class shouldn't be needed, I'm not sure there's any need for the `{docker,podman}_client_fixture` fixtures either, and the logic in this class can just go in the `ctr_client` fixture directly?

Reply:
We need at least two fixtures, one at session scope (for the availability testing) and one with a function scope (for skipping if necessary).
I just used a class because I needed to transfer 3 variables between fixtures (client, availability, message) and using a tuple was bloating the type hints.
I'm open to a refactoring of course if something can be simplified

Reply:
(although maybe my comment https://github.com/gabrieldemarmiesse/python-on-whales/pull/523#discussion_r1445024503 adds a reason for `docker_client` and `podman_client` fixtures, depending on how we solve that)

Reply:
> We need at least two fixtures, one at session scope (for the availability testing) and one with a function scope (for skipping if necessary).

I think you can use `pytest.skip()` in any fixture? https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-skip

Reply:
Sure but I wouldn't make sense to call skip in a function called once per test session

Reply:
If we want to skip all tests using a fixture that makes sense to be session-scoped then it very much does make sense to call `pytest.skip()` in it? Assuming that works (which I'm not certain of).